File: IOS/Sources/PeerProtocol.swift
import Foundation
import UIKit

struct PeerRemoteFile: Identifiable, Codable, Hashable {
    let id: String
    let name: String
    let size: Int64
    let mime: String
}

struct PeerInfo: Codable {
    let protocolName: String
    let deviceId: String
    let lobbyId: String
    let alias: String
    let platform: String
    let peerPort: Int
    let legacyPort: Int
    let lobbyActive: Bool
    let fileCount: Int
    let files: [PeerRemoteFile]
    let compatible: [String]?
    let capabilities: [String]?

    enum CodingKeys: String, CodingKey {
        case protocolName = "protocol"
        case deviceId, lobbyId, alias, platform, peerPort, legacyPort, lobbyActive, fileCount, files, compatible, capabilities
    }
}

struct PeerLobby: Identifiable, Hashable {
    var id: String { deviceId }
    let deviceId: String
    let lobbyId: String
    let alias: String
    let platform: String
    let host: String
    let peerPort: Int
    let legacyPort: Int
    let files: [PeerRemoteFile]
    let protocolVersion: String
    let capabilities: Set<String>

    var supportsOffers: Bool {
        protocolVersion == PeerProtocol.versionV2 && capabilities.contains(PeerProtocol.capOffers)
    }
}

enum PeerIdentity {
    private static let key = "FTPortalPeerDeviceId"

    static func deviceId() -> String {
        if let existing = UserDefaults.standard.string(forKey: key), !existing.isEmpty {
            return existing
        }
        let created = UUID().uuidString.replacingOccurrences(of: "-", with: "")
        UserDefaults.standard.set(created, forKey: key)
        return created
    }

    static func alias() -> String {
        let name = UIDevice.current.name.trimmingCharacters(in: .whitespacesAndNewlines)
        return name.isEmpty ? "iPhone / iPad" : name
    }
}

enum PeerProtocol {
    static let versionV1 = "ftportal/1"
    static let versionV2 = "ftportal/2"
    static let version = versionV2
    static let peerPort = 47_171

    static let infoPathV1 = "/api/ftportal/v1/info"
    static let downloadPrefixV1 = "/api/ftportal/v1/download/"
    static let infoPathV2 = "/api/ftportal/v2/info"
    static let offerPathV2 = "/api/ftportal/v2/offers"
    static let transferPrefixV2 = "/api/ftportal/v2/transfers/"

    // Compatibility aliases for the established direct-pull UI.
    static let infoPath = infoPathV1
    static let downloadPrefix = downloadPrefixV1

    static let capOffers = "offers"
    static let capAcceptDecline = "accept-decline"
    static let capBearerToken = "bearer-token"
    static let capVerificationCode = "verification-code"
    static let capabilitiesV2 = ["lobbies", capOffers, capAcceptDecline, capBearerToken, capVerificationCode]

    static func infoData(legacyPort: Int, version: String = versionV2) -> Data {
        let files = PortalStore.shared.all().map {
            PeerRemoteFile(id: $0.id, name: $0.name, size: $0.size, mime: $0.mime)
        }
        let deviceId = PeerIdentity.deviceId()
        let info = PeerInfo(
            protocolName: version,
            deviceId: deviceId,
            lobbyId: "lobby-\(deviceId)",
            alias: PeerIdentity.alias(),
            platform: "iOS",
            peerPort: peerPort,
            legacyPort: legacyPort,
            lobbyActive: !files.isEmpty,
            fileCount: files.count,
            files: files,
            compatible: version == versionV2 ? [versionV1] : nil,
            capabilities: version == versionV2 ? capabilitiesV2 : nil
        )
        return (try? JSONEncoder().encode(info)) ?? Data("{\"protocol\":\"\(version)\",\"lobbyActive\":false,\"files\":[]}".utf8)
    }
}

enum PeerDiscovery {
    static func discover() async -> [PeerLobby] {
        let ownAddresses = Set(PortalServer.localIPv4())
        let ownId = PeerIdentity.deviceId()
        var candidates = Set<String>()

        for address in ownAddresses {
            let parts = address.split(separator: ".")
            guard parts.count == 4 else { continue }
            let prefix = parts.prefix(3).joined(separator: ".")
            for last in 1...254 {
                let candidate = "\(prefix).\(last)"
                if !ownAddresses.contains(candidate) { candidates.insert(candidate) }
            }
        }
        if candidates.isEmpty { return [] }

        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 0.8
        configuration.timeoutIntervalForResource = 1.4
        configuration.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        let session = URLSession(configuration: configuration)
        defer { session.invalidateAndCancel() }

        let hosts = Array(candidates)
        var found: [String: PeerLobby] = [:]
        let batchSize = 32

        for start in stride(from: 0, to: hosts.count, by: batchSize) {
            let end = min(start + batchSize, hosts.count)
            let batch = hosts[start..<end]
            await withTaskGroup(of: PeerLobby?.self) { group in
                for host in batch {
                    group.addTask {
                        await probe(host: host, session: session, ownId: ownId)
                    }
                }
                for await lobby in group {
                    if let lobby, found[lobby.deviceId] == nil {
                        found[lobby.deviceId] = lobby
                    }
                }
            }
        }

        return found.values.sorted {
            $0.alias.localizedCaseInsensitiveCompare($1.alias) == .orderedAscending
        }
    }

    private static func probe(host: String, session: URLSession, ownId: String) async -> PeerLobby? {
        if let v2 = await probeVersion(
            host: host,
            path: PeerProtocol.infoPathV2,
            clientVersion: PeerProtocol.versionV2,
            session: session,
            ownId: ownId
        ) { return v2 }

        return await probeVersion(
            host: host,
            path: PeerProtocol.infoPathV1,
            clientVersion: PeerProtocol.versionV1,
            session: session,
            ownId: ownId
        )
    }

    private static func probeVersion(
        host: String,
        path: String,
        clientVersion: String,
        session: URLSession,
        ownId: String
    ) async -> PeerLobby? {
        guard let url = URL(string: "http://\(host):\(PeerProtocol.peerPort)\(path)") else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 0.8
        request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(clientVersion, forHTTPHeaderField: "X-FTPortal-Client")

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
            let info = try JSONDecoder().decode(PeerInfo.self, from: data)
            guard
                (info.protocolName == PeerProtocol.versionV2 || info.protocolName == PeerProtocol.versionV1),
                info.deviceId != ownId,
                info.lobbyActive,
                !info.files.isEmpty
            else { return nil }

            return PeerLobby(
                deviceId: info.deviceId,
                lobbyId: info.lobbyId,
                alias: info.alias,
                platform: info.platform,
                host: host,
                peerPort: (1...65_535).contains(info.peerPort) ? info.peerPort : PeerProtocol.peerPort,
                legacyPort: info.legacyPort,
                files: info.files,
                protocolVersion: info.protocolName,
                capabilities: Set(info.capabilities ?? [])
            )
        } catch {
            return nil
        }
    }
}

File: ANDROID/app/src/main/java/com/mrhakan/ftportal/PeerProtocol.kt
package com.mrhakan.ftportal

import android.content.Context
import android.os.Build
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger

data class PeerRemoteFile(
    val id: String,
    val name: String,
    val size: Long,
    val mime: String
)

data class PeerLobby(
    val deviceId: String,
    val lobbyId: String,
    val alias: String,
    val platform: String,
    val host: String,
    val peerPort: Int,
    val legacyPort: Int,
    val files: List<PeerRemoteFile>,
    val protocolVersion: String = PeerProtocol.VERSION_V1,
    val capabilities: Set<String> = emptySet()
) {
    val supportsOffers: Boolean
        get() = protocolVersion == PeerProtocol.VERSION_V2 && capabilities.contains(PeerProtocol.CAP_OFFERS)
}

object PeerIdentity {
    private const val PREFS = "ftportal_peer"
    private const val DEVICE_ID = "device_id"

    fun deviceId(context: Context): String {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val existing = prefs.getString(DEVICE_ID, null)
        if (!existing.isNullOrBlank()) return existing
        val created = UUID.randomUUID().toString().replace("-", "")
        prefs.edit().putString(DEVICE_ID, created).commit()
        return created
    }

    fun alias(): String {
        val manufacturer = Build.MANUFACTURER.orEmpty().trim()
        val model = Build.MODEL.orEmpty().trim()
        return listOf(manufacturer, model)
            .filter { it.isNotBlank() }
            .joinToString(" ")
            .ifBlank { "Android device" }
    }
}

object PeerProtocol {
    const val VERSION_V1 = "ftportal/1"
    const val VERSION_V2 = "ftportal/2"
    const val VERSION = VERSION_V2
    const val PEER_PORT = 47171

    const val INFO_PATH_V1 = "/api/ftportal/v1/info"
    const val DOWNLOAD_PREFIX_V1 = "/api/ftportal/v1/download/"
    const val INFO_PATH_V2 = "/api/ftportal/v2/info"
    const val OFFER_PATH_V2 = "/api/ftportal/v2/offers"
    const val TRANSFER_PREFIX_V2 = "/api/ftportal/v2/transfers/"

    // Compatibility aliases for the established pull flow.
    const val INFO_PATH = INFO_PATH_V1
    const val DOWNLOAD_PREFIX = DOWNLOAD_PREFIX_V1

    const val CAP_OFFERS = "offers"
    const val CAP_ACCEPT_DECLINE = "accept-decline"
    const val CAP_BEARER_TOKEN = "bearer-token"
    const val CAP_VERIFICATION_CODE = "verification-code"

    private val v2Capabilities = listOf(
        "lobbies",
        CAP_OFFERS,
        CAP_ACCEPT_DECLINE,
        CAP_BEARER_TOKEN,
        CAP_VERIFICATION_CODE
    )

    fun infoJson(context: Context, legacyPort: Int, version: String = VERSION_V2): String {
        val files = ShareRegistry.all()
        val deviceId = PeerIdentity.deviceId(context)
        val payload = JSONObject()
            .put("protocol", version)
            .put("deviceId", deviceId)
            .put("lobbyId", "lobby-$deviceId")
            .put("alias", PeerIdentity.alias())
            .put("platform", "Android")
            .put("peerPort", PEER_PORT)
            .put("legacyPort", legacyPort)
            .put("lobbyActive", files.isNotEmpty())
            .put("fileCount", files.size)

        if (version == VERSION_V2) {
            payload.put("compatible", JSONArray().put(VERSION_V1))
            payload.put("capabilities", JSONArray(v2Capabilities))
        }

        val manifest = JSONArray()
        files.forEach { file ->
            manifest.put(
                JSONObject()
                    .put("id", file.id)
                    .put("name", file.name)
                    .put("size", file.size)
                    .put("mime", file.mime)
            )
        }
        payload.put("files", manifest)
        return payload.toString()
    }

    fun parseLobby(host: String, body: String): PeerLobby? = runCatching {
        val json = JSONObject(body)
        val protocol = json.optString("protocol")
        if (protocol != VERSION_V2 && protocol != VERSION_V1) return@runCatching null
        if (!json.optBoolean("lobbyActive", false)) return@runCatching null

        val filesJson = json.optJSONArray("files") ?: JSONArray()
        val files = buildList {
            for (index in 0 until filesJson.length()) {
                val file = filesJson.optJSONObject(index) ?: continue
                val id = file.optString("id")
                val name = file.optString("name")
                if (id.isBlank() || name.isBlank()) continue
                add(
                    PeerRemoteFile(
                        id = id,
                        name = name,
                        size = file.optLong("size", -1L),
                        mime = file.optString("mime", "application/octet-stream")
                    )
                )
            }
        }
        if (files.isEmpty()) return@runCatching null

        val capabilitiesJson = json.optJSONArray("capabilities") ?: JSONArray()
        val capabilities = buildSet {
            for (index in 0 until capabilitiesJson.length()) {
                val value = capabilitiesJson.optString(index)
                if (value.isNotBlank()) add(value)
            }
        }
        val deviceId = json.getString("deviceId")
        PeerLobby(
            deviceId = deviceId,
            lobbyId = json.optString("lobbyId", "lobby-$deviceId"),
            alias = json.optString("alias", host),
            platform = json.optString("platform", "Unknown"),
            host = host,
            peerPort = json.optInt("peerPort", PEER_PORT).takeIf { it in 1..65535 } ?: PEER_PORT,
            legacyPort = json.optInt("legacyPort", 0),
            files = files,
            protocolVersion = protocol,
            capabilities = capabilities
        )
    }.getOrNull()
}

object PeerDiscovery {
    fun discover(context: Context, completion: (List<PeerLobby>) -> Unit) {
        val ownAddresses = NetworkUrls.ipv4Addresses().toSet()
        val candidates = linkedSetOf<String>()
        ownAddresses.forEach { address ->
            val parts = address.split('.')
            if (parts.size != 4) return@forEach
            val prefix = parts.take(3).joinToString(".")
            for (last in 1..254) {
                val candidate = "$prefix.$last"
                if (candidate !in ownAddresses) candidates += candidate
            }
        }

        if (candidates.isEmpty()) {
            completion(emptyList())
            return
        }

        val ownId = PeerIdentity.deviceId(context)
        val results = ConcurrentHashMap<String, PeerLobby>()
        val remaining = AtomicInteger(candidates.size)
        val pool = Executors.newFixedThreadPool(32)

        candidates.forEach { host ->
            pool.execute {
                try {
                    val lobby = probe(host)
                    if (lobby != null && lobby.deviceId != ownId) {
                        results.putIfAbsent(lobby.deviceId, lobby)
                    }
                } finally {
                    if (remaining.decrementAndGet() == 0) {
                        pool.shutdown()
                        completion(results.values.sortedBy { it.alias.lowercase() })
                    }
                }
            }
        }
    }

    private fun probe(host: String): PeerLobby? {
        val probes = listOf(
            PeerProtocol.INFO_PATH_V2 to PeerProtocol.VERSION_V2,
            PeerProtocol.INFO_PATH_V1 to PeerProtocol.VERSION_V1
        )
        for ((path, version) in probes) {
            val lobby = probeVersion(host, path, version)
            if (lobby != null) return lobby
        }
        return null
    }

    private fun probeVersion(host: String, path: String, version: String): PeerLobby? = runCatching {
        val connection = (URL("http://$host:${PeerProtocol.PEER_PORT}$path").openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 350
            readTimeout = 700
            instanceFollowRedirects = false
            useCaches = false
            setRequestProperty("Accept", "application/json")
            setRequestProperty("X-FTPortal-Client", version)
        }
        try {
            if (connection.responseCode != HttpURLConnection.HTTP_OK) return@runCatching null
            val body = connection.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
            PeerProtocol.parseLobby(host, body)
        } finally {
            connection.disconnect()
        }
    }.getOrNull()
}
